### PR TITLE
[Merged by Bors] - feature(LinearAlgebra/QuadraticForm/Basis): toBilinHom

### PR DESCRIPTION
**Pull request by mans0954**
For each basis `bm`, `QuadraticMap.toBilinHom` is the linear map that sends a quadratic map on a module `M` over `R`
to its associated symmetric bilinear map.

For free modules, `toBilinHom`  is the `toBilin` equivalent of `QuadraticMap.associatedHom`

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/QuadraticForm/Basis.lean [hunk 1]
```diff
@@ -60,4 +62,25 @@ theorem _root_.LinearMap.BilinMap.toQuadraticMap_surjective [Module.Free R M] :
...
+    obtain h | rfl | h := lt_trichotomy i j
+    · simp [h.ne, h, toBilin_apply, polar_add]
+    · simp [toBilin_apply]
+    · simp [h.ne', h.not_lt, toBilin_apply, polar_add]
+  map_smul' r Q := by
+    refine bm.ext fun i => bm.ext fun j => ?_
+    obtain h | rfl | h := lt_trichotomy i j
+    · simp [h.ne, h, toBilin_apply, polar_smul]
+    · simp [toBilin_apply]
+    · simp [h.ne', h.not_lt, toBilin_apply]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basis.lean:84]: Can you state these as standalone `simp` theorems first?

File: Mathlib/LinearAlgebra/QuadraticForm/Basis.lean [hunk 2]
```diff
@@ -60,4 +62,25 @@ theorem _root_.LinearMap.BilinMap.toQuadraticMap_surjective [Module.Free R M] :
   letI : LinearOrder ι := IsWellOrder.linearOrder WellOrderingRel
   exact ⟨_, toQuadraticMap_toBilin _ b⟩
 
+variable (S) [CommSemiring S] [Algebra S R]
+variable [Module S N] [IsScalarTower S R N]
+
+/-- For each basis `bm`, `toBilinHom` is the map that sends a quadratic map on a module `M` over `R`
+to its associated symmetric bilinear map.  As provided here, this has the structure of an
+`S`-linear map where `S` is a commutative subring of `R`. -/
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basis.lean:70]: I would just say "`toBilin` as a linear map"

File: Mathlib/LinearAlgebra/QuadraticForm/Basis.lean [hunk 3]
```diff
@@ -60,4 +62,25 @@ theorem _root_.LinearMap.BilinMap.toQuadraticMap_surjective [Module.Free R M] :
...
+noncomputable def toBilinHom (bm : Basis ι R M) : QuadraticMap R M N →ₗ[S] (BilinMap R M N) where
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basis.lean:71]:
```suggestion
@[simps]
noncomputable def toBilinHom (bm : Basis ι R M) : QuadraticMap R M N →ₗ[S] (BilinMap R M N) where
```

**Review by eric-wieser — APPROVED** [review]

File: Mathlib/LinearAlgebra/QuadraticForm/Basis.lean (hunk 1 above)
Comment (mans0954) [reply Mathlib/LinearAlgebra/QuadraticForm/Basis.lean]: @eric-wieser done - have I got the names right?

File: Mathlib/LinearAlgebra/QuadraticForm/Basis.lean [hunk 4]
```diff
@@ -60,4 +62,32 @@ theorem _root_.LinearMap.BilinMap.toQuadraticMap_surjective [Module.Free R M] :
...
+    (s • Q).toBilin bm = s • (Q.toBilin bm) := by
+  refine bm.ext fun i => bm.ext fun j => ?_
+  obtain h | rfl | h := lt_trichotomy i j
+  · simp [h.ne, h, toBilin_apply, polar_smul]
+  · simp [toBilin_apply]
+  · simp [h.ne', h.not_lt, toBilin_apply]
+
+/-- `QuadraticMap.toBilin` as an S-linear map -/
+@[simps]
+noncomputable def toBilinHom (bm : Basis ι R M) : QuadraticMap R M N →ₗ[S] (BilinMap R M N) where
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basis.lean:88]: nit:
```suggestion
noncomputable def toBilinHom (bm : Basis ι R M) : QuadraticMap R M N →ₗ[S] BilinMap R M N where
```

File: Mathlib/LinearAlgebra/QuadraticForm/Basis.lean [hunk 5]
```diff
@@ -60,4 +62,32 @@ theorem _root_.LinearMap.BilinMap.toQuadraticMap_surjective [Module.Free R M] :
...
+  · simp [h.ne, h, toBilin_apply, polar_add]
+  · simp [toBilin_apply]
+  · simp [h.ne', h.not_lt, toBilin_apply, polar_add]
+
+variable (S) [CommSemiring S] [Algebra S R]
+variable [Module S N] [IsScalarTower S R N]
+
+@[simp]
+lemma smul_toBilin (bm : Basis ι R M) (s : S) (Q : QuadraticMap R M N) :
+    (s • Q).toBilin bm = s • (Q.toBilin bm) := by
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/QuadraticForm/Basis.lean:79]: nit:
```suggestion
    (s • Q).toBilin bm = s • Q.toBilin bm := by
```

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

**Pull request closed without merging**